test/extensions/common/aws/metadata_fetcher_test.cc

```diff
@@ -0,0 +1,301 @@
+  MockMetadataReceiver receiver;
+  EXPECT_CALL(receiver, onMetadataSuccess(testing::_)).Times(0);
+  EXPECT_CALL(receiver,
+              onMetadataError(MetadataFetcher::MetadataReceiver::Failure::InvalidMetadata));
+
+  // Act
+  fetcher_->fetch(message, parent_span_, receiver);
+}
+
+TEST_F(MetadataFetcherTest, TestHttpFailure) {
+  // Setup
+  setupFetcher();
+  Http::RequestMessageImpl message;
+
+  MockUpstream mock_result(mock_factory_ctx_.cluster_manager_,
+                           Http::AsyncClient::FailureReason::Reset);
+  MockMetadataReceiver receiver;
+  EXPECT_CALL(receiver, onMetadataSuccess(testing::_)).Times(0);
+  EXPECT_CALL(receiver, onMetadataError(MetadataFetcher::MetadataReceiver::Failure::Network));
+
+  // Act
+  fetcher_->fetch(message, parent_span_, receiver);
+}
+
+TEST_F(MetadataFetcherTest, TestWithValidCluster) {
+  // Setup with thread local cluster
+  Http::RequestMessageImpl message;
+  message.headers().setScheme(Http::Headers::get().SchemeValues.Http);
+  message.headers().setMethod(Http::Headers::get().MethodValues.Get);
+  message.headers().setHost("169.254.170.2:80");
+  message.headers().setPath("/v2/credentials/c68caeb5-ef71-4914-8170-111111111111");
+
+  NiceMock<Upstream::MockThreadLocalCluster> cluster_;
+  fetcher_ = MetadataFetcher::create(mock_factory_ctx_.cluster_manager_, "cluster_name");
+  EXPECT_CALL(mock_factory_ctx_.cluster_manager_, getThreadLocalCluster(_))
+      .WillOnce(Return(&cluster_));
+
+  MockUpstream mock_result(mock_factory_ctx_.cluster_manager_, "200", "not_empty");
+  MockMetadataReceiver receiver;
+  EXPECT_CALL(receiver, onMetadataError(testing::_)).Times(0);
+
+  // Act
+  fetcher_->fetch(message, parent_span_, receiver);
+}
+
+TEST_F(MetadataFetcherTest, TestClusterNotFound) {
+  // Setup without thread local cluster
+  fetcher_ = MetadataFetcher::create(mock_factory_ctx_.cluster_manager_, "cluster_name");
+  Http::RequestMessageImpl message;
+  MockMetadataReceiver receiver;
+
+  EXPECT_CALL(mock_factory_ctx_.cluster_manager_, getThreadLocalCluster(_))
+      .WillOnce(Return(nullptr));
+  EXPECT_CALL(receiver, onMetadataSuccess(testing::_)).Times(0);
+  EXPECT_CALL(receiver, onMetadataError(MetadataFetcher::MetadataReceiver::Failure::MissingConfig));
+
+  // Act
+  fetcher_->fetch(message, parent_span_, receiver);
+}
+
+TEST_F(MetadataFetcherTest, TestCancel) {
+  // Setup
+  setupFetcher();
+  Http::RequestMessageImpl message;
+  Http::MockAsyncClientRequest request(
+      &(mock_factory_ctx_.cluster_manager_.thread_local_cluster_.async_client_));
+  MockUpstream mock_result(mock_factory_ctx_.cluster_manager_, &request);
+  MockMetadataReceiver receiver;
+  EXPECT_CALL(request, cancel());
+  EXPECT_CALL(receiver, onMetadataSuccess(testing::_)).Times(0);
+  EXPECT_CALL(receiver, onMetadataError(testing::_)).Times(0);
+
+  // Act
+  fetcher_->fetch(message, parent_span_, receiver);
+  // Proper cancel
+  fetcher_->cancel();
```

Comment:
Would be good to `Mock::VerifyAndClearExpectations(&request)` and receiver here, so that you're testing that "proper cancel" actually did it, and that "re-entrant cancel" did nothing.

Also, make the comment "re-entrant cancel should do nothing" to be extra clear. :)